Index: src/tools/msvc/Install.pm
===================================================================
--- src/tools/msvc/Install.pm	(revision 404)
+++ src/tools/msvc/Install.pm	(working copy)
@@ -339,7 +339,7 @@
 
             # Special case for contrib/spi
             $flist =
-"README.spi autoinc.example insert_username.example moddatetime.example refint.example timetravel.example"
+"autoinc.example insert_username.example moddatetime.example refint.example timetravel.example"
               if ($d eq 'spi');
             foreach my $f (split /\s+/,$flist)
             {
