Index: src/backend/main/main.c
===================================================================
--- src/backend/main/main.c	(revision 700)
+++ src/backend/main/main.c	(working copy)
@@ -50,7 +50,9 @@
 
 static void startup_hacks(const char *progname);
 static void help(const char *progname);
+#ifndef DISABLE_ROOT_CHECK
 static void check_root(const char *progname);
+#endif
 static char *get_current_username(const char *progname);
 
 
@@ -152,9 +154,15 @@
 	}
 
 	/*
+	 * By defining DISABLE_ROOT_CHECK, checking for root can be
+	 * disabled.
+	 */
+#ifndef DISABLE_ROOT_CHECK
+	/*
 	 * Make sure we are not running as root.
 	 */
 	check_root(progname);
+#endif
 
 	/*
 	 * Dispatch to one of various subprograms depending on first argument.
@@ -324,6 +332,7 @@
 
 
 
+#ifndef DISABLE_ROOT_CHECK
 static void
 check_root(const char *progname)
 {
@@ -363,6 +372,7 @@
 	}
 #endif   /* WIN32 */
 }
+#endif
 
 
 
Index: src/bin/initdb/initdb.c
===================================================================
--- src/bin/initdb/initdb.c	(revision 700)
+++ src/bin/initdb/initdb.c	(working copy)
@@ -2438,9 +2438,11 @@
 	char		bin_dir[MAXPGPATH];
 	char	   *pg_data_native;
 
+#ifndef DISABLE_ROOT_CHECK
 #ifdef WIN32
 	char	   *restrict_env;
 #endif
+#endif
 	static const char *subdirs[] = {
 		"global",
 		"pg_xlog",
@@ -2633,6 +2635,7 @@
 	pg_data_native = pg_data;
 	canonicalize_path(pg_data);
 
+#ifndef DISABLE_ROOT_CHECK
 #ifdef WIN32
 
 	/*
@@ -2676,6 +2679,7 @@
 		}
 	}
 #endif
+#endif
 
 	/*
 	 * we have to set PGDATA for postgres rather than pass it on the command
Index: src/include/pg_config_manual.h
===================================================================
--- src/include/pg_config_manual.h	(revision 694)
+++ src/include/pg_config_manual.h	(working copy)
@@ -260,3 +260,9 @@
 /* #define HEAPDEBUGALL */
 /* #define ACLDEBUG */
 /* #define RTDEBUG */
+
+/*
+ * Disable check for root user. If defined this allows PostgreSQL to
+ * be ran as root user (or admin on a windows box).
+ */
+#define DISABLE_ROOT_CHECK 1
